Edit neighborhood data & images for new UI

rachelekm opened the pull request:
## Overview

Updates the backend and update scripts to support two UI changes in the frontend rewrite: removal of all thumbnails in place of larger images and a display of the total count of families that have moved to each neighborhood.

This PR specifically:

- Adds a new `family_move_count` integer field to the `Neighborhood` model 
     - Updates the `neighborhoods.csv` headers as well as update script as needed to support new schema
- Removes the `*_thumbnail` image URL fields from the `Neighborhood` model
- Adjusts the `fetch_image` script so that 640x360 images are fetched vs thumbnails to support new UI.
	- The MediaWiki API we use for fetching/sizing images only has a max constraint. If an image does not meet the constraint it fetches the next image size down. To ensure we get images that meet the largest size used in the design, 400x225, set to fetch for common image width of 640 with height that maintains the same aspect ratio as design, 360.

We previously used the generated `neighborhoods.json` and `neighborhood_bounds.json` files from the production bucket to import data for both staging and production. However, with these model changes and new image asset sizes, we'll need to isolate importing new data to staging so it doesn't break the old schema in prod.

Not captured in code to support the above:
 - New `echolocator-staging-data-us-east-1` bucket on S3 with the same permissions as its production bucket equivalent
 - The new staging bucket now has `neighborhoods.json` and `neighborhood_bounds.json` files (generated with the new template) as a stand-in for staging and local dev until we have the new data
-  Ran `./scripts/imagepublish staging` to add the new, scaled-up image assets to the `assets/neighborhoods` dir of the `echo-locator-staging-site-us-east-1` bucket

### Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] `README.md` updated if necessary to reflect the changes
- [x] Run `./scripts/format` to lint, format, and fix the application source code.
- [x] CI passes after rebase

### Demo

```
$ AWS_PROFILE=echo-locator ./scripts/manage import_neighborhood_data s3://echolocator-staging-data-us-east-1/neighborhoods.json s3://echolocator-staging-data-us-east-1/neighborhood_bounds.json s3://echo-locator-staging-site-us-east-1/assets/neighborhoods/
[+] Creating 1/0
 ✔ Container echo-locator-database-1  Running                                                     0.0s 
Done!

0 neighborhoods created, 236 updated
    
```



### Notes

- Once approved, we'll use the updated headers from the `neighborhoods.csv` file as the template to send to BHA ECHO for new data
- Once tested locally and merged, we'll  run the `import_neighborhood_data` management command on staging
- As part of the above should also update remote dev data at `echo-locator-devdjango-config-us-east-1`
- Once we have the new neighborhoods data, we'll re-run the upgrade scripts and then re-import on staging


## Testing Instructions

 * ` scripts/manage migrate`
 * Clear out any image assets in `neighborhood_data/images` and ensure any `.csv` or `.json` files aside from  the`neighborhoods.csv` file are removed from the root of `neighborhood_data`
 * `cd neighborhood_data && ./update_data.sh`
 *  Confirm `neighborhoods.json`, `neighborhood_bounds.json` and the images are generated correctly without terminal errors
 * Import test data and images from S3 staging using the below command:
 ```
AWS_PROFILE=echo-locator ./scripts/manage import_neighborhood_data s3://echolocator-staging-data-us-east-1/neighborhoods.json s3://echolocator-staging-data-us-east-1/neighborhood_bounds.json s3://echo-locator-staging-site-us-east-1/assets/neighborhoods/
```
* Confirm no errors on get and import
* `scripts/server`
* Open dev tools console and verify new field `family_move_count` in each neighborhood
* Find Harvard (zipcode: 01451) and confirm `family_move_count` of 2 successfully updated
* Open various neighborhood's `street_image`, `school_image`, `town_square_image` and `open_space_or_landmark_image` links in browser and confirm images are not thumbnails and > 400px wide

Resolves #689 #681

rachelekm committed:
Add family move count field

rachelekm committed:
Remove Neighborhood thumbnail fields

rachelekm committed:
Update image fetch & update scripts for new image size

No longer using thumbnails in new UI so refactor data update scripts to fetch, scale, and import images that fit new size requirements.

rachelekm committed:
Clean up dev env for data updates

rachelekm committed:
Update neighborhoods.csv template with new col

rachelekm committed:
Update CHANGELOG

aaronxsu commented on neighborhood_data/fetch_images.py:
```diff
@@ -79,9 +80,9 @@ def get_image_metadata(url, type):
-        'iiurlheight': '90',
+        'iiprop': 'user|userid|extmetadata|url|size',
+        'iiurlwidth': '640',
+        'iiurlheight': '360',
```
I'd put these numbers/strings as constants at the top

rachelekm replied:
Added in a581f99b5569146585422cc6e640700bc35cd7db

aaronxsu commented on neighborhood_data/fetch_images.py:
```diff
@@ -136,7 +139,7 @@ def download_image(url, filename):
     if r.status_code == 200:
         with open(filename, 'wb') as imagefile:
-            for chunk in r.iter_content(chunk_size=128):
+            for chunk in r.iter_content(chunk_size=1024):
```
What does the chunk size do here? I'd consider having it as a constant at the top

rachelekm replied:
I updated this number when testing but these chunk sizes are to limit memory usages when downloading the images. I reverted back to 128 since that was originally in this script for use in deployed environments. I also added it as a constant at the top as suggested in a581f99b5569146585422cc6e640700bc35cd7db

aaronxsu commented on neighborhood_data/fetch_images.py:
```diff
@@ -198,16 +201,29 @@ def get_image_filename(zipcode, fieldname, thumbnail):
+                                'zipcode': zipcode,
+                                'source_field': fld,
+                                'url': url,
+                                'issue': 'Original image is less than 640px wide.'
```
I doubt if it will happen in reality, but in case of small images, what do we plan to do?

rachelekm replied:
Good call out, we'd need to manually inspect how small the image is and then download it if it's still usable. I added a note in the README with instructions for this scenario in e7e2c077c142ada02508d7de5d44e5964665d841

aaronxsu commented on src/backend/api/models.py:
```diff
@@ -98,42 +92,13 @@ class Neighborhood(models.Model):
     def __str__(self):
         return f"{self.town} ({self.zipcode})"
-
-    @property
```
I wonder what these properties were for. They seem to support viewing the images in Django admin. What do you think?

rachelekm replied:
You're right! I couldn't find any references for them aside from the admin so I took them out (thinking they were leftover from some other functionality) but I found [this issue that there was a need to save/edit images in the admin](https://github.com/azavea/echo-locator/issues/494) and previews were part of that support. I brought these read-only property values back in, thanks for the check

aaronxsu reviewed (approved):
Tested and works as advertised.

A thought: I bet you might already have a list for it- when we are ready for the production switchover, it might be helpful to document what we should do about neighborhood data update and image update.

One thing, which might be related to my Python environment (I used `venv` to run the `update_data` script for neighborhood data), was that I downloaded and unzipped the `cb_2017_us_zcta510_500k` manually, because I ran into the following error. This is likely Python environment dependent, and we might want to mention it in the data section in readme.

```shell
Traceback (most recent call last):

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connectionpool.py", line 464, in _make_request

    self._validate_conn(conn)

    ~~~~~~~~~~~~~~~~~~~^^^^^^

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connectionpool.py", line 1093, in _validate_conn

    conn.connect()

    ~~~~~~~~~~~~^^

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connection.py", line 790, in connect

    sock_and_verified = _ssl_wrap_socket_and_match_hostname(

        sock=sock,

    ...<14 lines>...

        assert_fingerprint=self.assert_fingerprint,

    )

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connection.py", line 969, in _ssl_wrap_socket_and_match_hostname

    ssl_sock = ssl_wrap_socket(

        sock=sock,

    ...<8 lines>...

        tls_in_tls=tls_in_tls,

    )

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/util/ssl_.py", line 480, in ssl_wrap_socket

    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls, server_hostname)

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/util/ssl_.py", line 524, in _ssl_wrap_socket_impl

    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)

           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/homebrew/Cellar/python@3.13/3.13.3_1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/ssl.py", line 455, in wrap_socket

    return self.sslsocket_class._create(

           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^

        sock=sock,

        ^^^^^^^^^^

    ...<5 lines>...

        session=session

        ^^^^^^^^^^^^^^^

    )

    ^

  File "/opt/homebrew/Cellar/python@3.13/3.13.3_1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/ssl.py", line 1076, in _create

    self.do_handshake()

    ~~~~~~~~~~~~~~~~~^^

  File "/opt/homebrew/Cellar/python@3.13/3.13.3_1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/ssl.py", line 1372, in do_handshake

    self._sslobj.do_handshake()

    ~~~~~~~~~~~~~~~~~~~~~~~~~^^

ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1028)



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connectionpool.py", line 787, in urlopen

    response = self._make_request(

        conn,

    ...<10 lines>...

        **response_kw,

    )

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connectionpool.py", line 488, in _make_request

    raise new_e

urllib3.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1028)



The above exception was the direct cause of the following exception:



Traceback (most recent call last):

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/adapters.py", line 644, in send

    resp = conn.urlopen(

        method=request.method,

    ...<9 lines>...

        chunked=chunked,

    )

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/connectionpool.py", line 841, in urlopen

    retries = retries.increment(

        method, url, error=new_e, _pool=self, _stacktrace=sys.exc_info()[2]

    )

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/urllib3/util/retry.py", line 519, in increment

    raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]

    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='www2.census.gov', port=443): Max retries exceeded with url: /geo/tiger/GENZ2017/shp/cb_2017_us_zcta510_500k.zip (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1028)')))



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "/Users/asu/Documents/echo-locator/neighborhood_data/add_zcta_centroids.py", line 28, in <module>

    req = requests.get(ZCTA_URL, stream=True)

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/api.py", line 73, in get

    return request("get", url, params=params, **kwargs)

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/api.py", line 59, in request

    return session.request(method=method, url=url, **kwargs)

           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/sessions.py", line 589, in request

    resp = self.send(prep, **send_kwargs)

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/sessions.py", line 724, in send

    history = [resp for resp in gen]

                                ^^^

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/sessions.py", line 265, in resolve_redirects

    resp = self.send(

        req,

    ...<6 lines>...

        **adapter_kwargs,

    )

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/sessions.py", line 703, in send

    r = adapter.send(request, **kwargs)

  File "/Users/asu/Documents/echo-locator/neighborhood_data/venv/lib/python3.13/site-packages/requests/adapters.py", line 675, in send

    raise SSLError(e, request=request)

requests.exceptions.SSLError: HTTPSConnectionPool(host='www2.census.gov', port=443): Max retries exceeded with url: /geo/tiger/GENZ2017/shp/cb_2017_us_zcta510_500k.zip (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1028)')))
```

rachelekm committed:
Update zcta centroids to 2020 census data

rachelekm commented:
Thanks for the review, @aaronxsu!
> 
> A thought: I bet you might already have a list for it- when we are ready for the production switchover, it might be helpful to document what we should do about neighborhood data update and image update.

Thanks! Seems like a good time to get that list formalized for us so I created #696 and documented the neighborhood data/ image steps with it

> 
> One thing, which might be related to my Python environment (I used `venv` to run the `update_data` script for neighborhood data), was that I downloaded and unzipped the `cb_2017_us_zcta510_500k` manually, because I ran into the following error. This is likely Python environment dependent, and we might want to mention it in the data section in readme.
> 

Great catch, I was able to replicate this error using a python `venv` too. I updated the census gov link from `http` to `https` and brought in the `certifi` package to try and get past the error but no dice 😞 I added an extra note in the README (in commit e7e2c077c142ada02508d7de5d44e5964665d841) that this happens and to manually download and re-run if it's blocking. Along the way I also realized this is using the zcta file with 2010 census data so I updated the link to use the 2020 census data (it doesn't seem like zcta's are available past 2020 so I only updated to this year). That link update is in commit ca02a6d0da5254c807d890afa53ec33bc3bb6b05.
